liferay-echo/liferay-portal
# [TS] LPS-170629

Motivation
=======
This is a solution to [LPS-170629](https://liferay.atlassian.net/browse/LPS-170629).
    
Some arguments of the StructuredContent API, like `relatedContents`, are readOnly, so one cannot specify related contents when adding or modifying something like a web content. 
    
A review of which arguments of the StructuredContent was done in [LPS-171394](https://liferay.atlassian.net/browse/LPS-171394) with the conclusion to modify:
    -  `relatedContents`
    -  `structuredContentFolderId`
    
Proposed Solution
========
Remove the `readOnly: true` specification for both `relatedContents` and `structuredContentFolderId` within `StructuredContent`.
    
Since `RelatedContent` has its own internal properties (`contentType`, `id`, and `title`), the first two (`contentType` and `id`) need to have the `readOnly: true` attribute removed, too.
    
Finally, when creating the JournalArticle from the StructuredContent information in StructuredContentResourceImpl.java`, the `RelatedContents` and the `structuredContentFolderId` need to be taken into account.
  
(Note: A `RelatedContent` has a `contentType` that needs to be translated into a `className`. For example `StructuredContent` is translated as `JournalArticle`. This map is explicitly built with `_assetTypeTypeToClassNames` similarly to what is done in `TaxonomyVocabularyResourceImpl.java`.)
  
The tests for StructuredContent need to be adapted now that `relatedContents` and `structuredContentFolderId` can be modified. To test the `relatedContents` I added a `blogsEntry`. For the `structuredContentFolderId` I could use the existing `JournalFolder`'s in the involved test classes.

cc @ruben-pulido 
    
Steps to Verify
========
See reproducing steps in [LPS-170629](https://liferay.atlassian.net/browse/LPS-170629).

Checklist (optional)
========

CODE:

COMMITS:

PR TITLE & DESCRIPTION:
- [x] My PR description includes the steps to verify the change (points to LPS)
